multicluster/controllers/multicluster/memberclusterannounce_controller.go

```diff
@@ -18,45 +18,275 @@ package multicluster
+							condition.Status = data.leaderStatus.electedLocalAsLeaderCluster
+							condition.Message = data.leaderStatus.message
+							condition.LastTransitionTime = metav1.Now()
+							condition.Reason = data.leaderStatus.reason
+						}
+					}
+				}
+			}
+		} else {
+			// Member has not updated MemberClusterStatus in the last 3 intervals, assume it is disconnected
+			for index := range status.Conditions {
+				condition := &status.Conditions[index]
+				switch condition.Type {
+				case multiclusterv1alpha1.ClusterReady:
+					{
+						if condition.Status != v1.ConditionFalse {
+							condition.Status = v1.ConditionFalse
+							condition.LastTransitionTime = metav1.Now()
+							condition.Message = fmt.Sprintf("No MemberClusterAnnounce update after %v", data.lastUpdateTime)
+							condition.Reason = ReasonDisconnected
+						}
+					}
+				case multiclusterv1alpha1.ClusterWatchesLocalResources:
+					{
+						if condition.Status != v1.ConditionFalse || condition.Reason != ReasonDisconnected {
+							condition.Status = v1.ConditionFalse
+							condition.Message = fmt.Sprintf("No MemberClusterAnnounce update after %v", data.lastUpdateTime)
+							condition.LastTransitionTime = metav1.Now()
+							condition.Reason = ReasonDisconnected
+						}
+					}
+				}
+			}
+		}
+	}
+}
+
+/******************************* MemberClusterStatusManager methods *******************************/
+
+func (r *MemberClusterAnnounceReconciler) AddMember(MemberId common.ClusterID) {
+	defer r.mapLock.Unlock()
+	r.mapLock.Lock()
+	if _, ok := r.memberStatus[MemberId]; ok {
+		// already present
+		return
+	}
+
+	conditions := make([]multiclusterv1alpha1.ClusterCondition, 0, 2)
+	conditions = append(conditions, multiclusterv1alpha1.ClusterCondition{
+		Type:               multiclusterv1alpha1.ClusterReady,
+		Status:             v1.ConditionUnknown,
+		LastTransitionTime: metav1.Now(),
+		Message:            "Member created",
+		Reason:             ReasonNeverConnected,
+	})
+	conditions = append(conditions, multiclusterv1alpha1.ClusterCondition{
+		Type:               multiclusterv1alpha1.ClusterWatchesLocalResources,
+		Status:             v1.ConditionFalse,
+		LastTransitionTime: metav1.Now(),
+		Message:            "Member created",
+		Reason:             ReasonNeverConnected,
+	})
+
+	r.memberStatus[MemberId] = &multiclusterv1alpha1.ClusterStatus{ClusterID: string(MemberId),
+		Conditions: conditions}
+
+	r.timerData[MemberId] = &timerData{connected: false, lastUpdateTime: time.Time{}}
+}
+
+func (r *MemberClusterAnnounceReconciler) RemoveMember(MemberId common.ClusterID) {
+	defer r.mapLock.Unlock()
+	r.mapLock.Lock()
+
+	if _, ok := r.memberStatus[MemberId]; ok {
+		delete(r.memberStatus, MemberId)
+		return
+	}
+
+	if _, ok := r.timerData[MemberId]; ok {
+		delete(r.timerData, MemberId)
+		return
+	}
+}
+
+func (r *MemberClusterAnnounceReconciler) GetMemberClusterStatues() []multiclusterv1alpha1.ClusterStatus {
```

Comment:
GetMemberClusterStatues -> GetMemberClusterStatuses

Reply:
oops, thanks for catching it. Fixing it.